### DISCUSSION TITLE
solver to solve Neighborhood

### DISCUSSION THREAD

**Discussion by C8504** [google/or-tools#3291] [CP-SAT questions] [answered]
Hi experts,

Could I set parameters and solver(use scip to solve neighborhoods) for lns subsolver in CP-SAT?

Currently I guess neighborhoods are solved by CP-SAT as well.

John

**Comment by lperron** [comment]
1) SCIP is worse that CP-SAT on pure integer problems
2) SCIP only solves a subset of the problems described in the CP-SAT proto

So why would I do that ?

**Reply by C8504** [reply]
Thanks for reply, lperron.
Actually, I wanted to set params for internal CP-SAT which solved neighborhoods for lns sub solver. For example time limit, threads, and so many SAT params.
Furthermore, I thought if it is possible to use third_party solver, scip or gurobi to solve the neighborhoods.

John

**Comment by lperron** [comment, marked as answer]
same answer. Any MIP solver will be very limited.

There is an exception though, you can enable max_hs through a command line
flag, which is a bender's decomposition approach (linearize part of the
model, optimize with a MIP solver, aka SCIP, inject the optimal solution
into the SAT solver, get an optimal solution, or a nogood (assumption),
inject the nogood in the MIP model). Unfortunately, it helps very rarely,
and degrades performance a lot of time. So it is only in experimental mode
now.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Reply by C8504** [reply]
ok, let's forget mip solver here. :-)

Could I set params for  lns subsolver(CP-SAT solver)?

John

**Reply by lperron** [reply]
yes. in 9.4.

You can select which workers to use, and overwrite/create params for workers.

See the complete section in the parameters: https://github.com/google/or-tools/blob/33fd1f910d0a9bb1aa922ff76aa5d502e3e1333f/ortools/sat/sat_parameters.proto#L534

**Reply by C8504** [reply]
thanks
